pkg/scheduler/extender/extender.go

```diff
@@ -329,6 +363,81 @@ func (e *Extender) filter(ctx context.Context, nodes []coreV1.Node, volumes []*g
 	return matchedNodes, failedNodesMap, err
 }
 
+func (e *Extender) score(nodes []coreV1.Node) ([]schedulerapi.HostPriority, error) {
+	ll := e.logger.WithFields(logrus.Fields{
+		"method": "score",
+	})
+
+	var volumeList = &volcrd.VolumeList{}
+	if err := e.k8sClient.ReadList(context.Background(), volumeList); err != nil {
+		err = fmt.Errorf("unable to read AvailableCapacity list: %v", err)
+		return nil, err
+	}
+	ll.Debugf("Got %d volumes", len(volumeList.Items))
+
+	nodeMapping := nodeVolumeCountMapping(volumeList)
+	sortedNoded := getSortedNodes(nodeMapping)
+	ll.Debugf("sorted volumes %v", sortedNoded)
+
+	priorityFromVolumes := nodePrioritize(sortedNoded)
+
+	ll.Debugf("nodes were ranked by their volumes %+v", priorityFromVolumes)
+	hostPriority := make([]schedulerapi.HostPriority, 0, len(nodes))
+	for _, node := range nodes {
+		// set the highest priority if node doesn't have any volumes
+		rank := math.MaxInt16
+		if r, ok := priorityFromVolumes[string(node.GetUID())]; ok {
+			rank = r
```

Comment:
Does rank mean just a number of PV? Shouldn't it be calculated depending on other node's values? Will high rank cause issue with scores from CPU/memory and other extenders?

Reply:
It's based on the position in the sorted slice of nodes based on volume count.
Apparently, scoring is used only with extenders. So other extenders may affect the score of one of the nodes.
But I don't see how it could be an issue - scoring runs on the already filtered list of nodes and other extenders can have higher scores.

Reply:
Let's say you have _**one node with 100 volumes**_ and **_second with 1 volume only_**. Final scores will be 0 and 1 correspondingly. Other extenders will simply over score node with 1 volume I guess. An vice versa - when you have no volumes on the node(s) _MaxInt16_ score will impact placement decisions of others.

Reply:
My intent was to force nodes without volume to have one. I can rewrite prioritize to make it more aligned on "something".
But there is no guarantee that other plugins won't over score us :)

Reply:
The original intent was to have even distribution of the volumes across the nodes (vSAN uses term _balanced_). You can calсulate rank base on the following formula: _rank of node X = max number of volumes - number of volume on node X_. For example, if you have 4 nodes with 8, 7, 7 and 5 volumes, final ranks will be 8-8=0, 8-7=1, 8-7=1, 8-5=3. In this case we will prioritize the last node clearly.

Reply:
Fixed.

Reply:
You use total volume count for all nodes, but you need to find max number of volumes on a single node across nodes list from _extenderArgs.Nodes.Items_.

Reply:
As a future improvement, we can  consider aligning priorities to MaxNodeScore. But it's a little bit confusing which value to use - [16.4](https://github.com/kubernetes/kubernetes/blob/224be7bdce5a9dd0c2fd0d46b83865648e2fe0ba/pkg/scheduler/framework/v1alpha1/interface.go#L74) - 10, [master](https://github.com/kubernetes/kubernetes/blob/1a66eb7b8aba8d5e66d855d6b8b31fdeab648778/pkg/scheduler/framework/v1alpha1/interface.go#L86) - 100.

Reply:
Yep, right now I'm getting the max number of volumes on a single node, not all.